airbyte-cdk/bulk/toolkits/load-s3/src/main/kotlin/io/airbyte/cdk/load/file/s3/S3MultipartUpload.kt

```diff
@@ -80,57 +86,56 @@ class S3MultipartUpload<T : OutputStream>(
     }
 
     inner class UploadStream : OutputStream() {
-        override fun close() = runBlocking {
-            if (closeOnce.setOnce()) {
-                workQueue.send { workQueue.close() }
+        override fun close() {
+            if (isClosed.setOnce()) {
+                partQueue.close()
             }
         }
 
-        override fun flush() = runBlocking { workQueue.send { wrappingBuffer.flush() } }
+        override fun flush() = wrappingBuffer.flush()
 
-        override fun write(b: Int) = runBlocking {
-            workQueue.send {
```

Comment:
help me out while trying to understand where the previous bug was - is it here?

Reply:
Yes, specifically it was paying the cost per byte of some combination of
* de/allocating this lambda passed through send
* de/allocating the hidden lambda due to `send` being a suspend fun
* whatever additional overhead is imposed by making interacting with the channel thread-safe

Reply:
the `write(b: ByteArray)` was also more costly than it needed to be, showing up as a slight perf hit on json/avro. But the big one was CSV.